### PR TITLE
Support the List/Tuple Construct/Unpack operation for TorchModuleGraph

### PR DESCRIPTION
**Pull request by zheng-ningxin**
Support the list/tuple construct/unpack operation for the TorchModuleGraph.
Fix the bug mentioned in https://github.com/microsoft/nni/issues/2581.

In the original version, we take the list/tuple construct/unpack operation nodes as unimportant nodes and merge them with the adjacent important nodes. However, merging the unpack nodes will lead to a graph construct error.

### CODE REVIEW

File: src/sdk/pynni/nni/_graph_utils.py [hunk 1]
```diff
@@ -199,6 +203,8 @@ def __init__(self, name, unique_name, node_type, op_type, node_cpps, inputs=None
             All the inputs of this node, each element is debugName of one input
         outputs: list of str
             All the outputs of this node, each element is debugName of one output
+        key_node: torch._C.Node
+            The key node of this NodePyGroup.
```
Comment (QuanluZhang) [inline src/sdk/pynni/nni/_graph_utils.py:207]: what is the meaning of key node?

File: src/sdk/pynni/nni/_graph_utils.py [hunk 2]
```diff
@@ -510,6 +518,65 @@ def _build_index(self, nodes_op):
...
+        """
+        Judge if a cpp node is a key function node.
+        If so, we should not merge this node into the
+        adjacent node.
+        """
+        if node_cpp.kind().startswith('aten::'):
+            # the nodes that start with 'aten' are key function
+            # nodes
+            return True
+        if node_cpp.kind() in [LIST_UNPACK_KIND, TUPLE_UNPACK_KIND]:
```
Comment (QuanluZhang) [inline src/sdk/pynni/nni/_graph_utils.py:531]: not construct type here?

File: src/sdk/pynni/nni/_graph_utils.py [hunk 3]
```diff
@@ -510,6 +518,65 @@ def _build_index(self, nodes_op):
...
+            # We cannot merge the List/Tuple
+            # Construct/Unpack func into other nodes, else it
+            # may lead to a graph construction error.
```
Comment (QuanluZhang) [inline src/sdk/pynni/nni/_graph_utils.py:534]: what is this error like?

File: src/sdk/pynni/nni/_graph_utils.py [hunk 4]
```diff
@@ -510,6 +518,65 @@ def _build_index(self, nodes_op):
...
+        the graph. Note: this function will change the
+        graph structure.
+        """
+        if hasattr(self, 'unpacked'):
+            # if already unpacked the tuple/list manually
+            return
+        for node in self.nodes_py.nodes_op:
+            if node.op_type in [TUPLE_UNPACK_KIND, LIST_UNPACK_KIND]:
+                unpack_cpp = node.key_node
+                last_cpp = list(unpack_cpp.inputs())[0].node()
```
Comment (QuanluZhang) [inline src/sdk/pynni/nni/_graph_utils.py:551]: index is 0, why call it `last`?

File: src/sdk/pynni/nni/_graph_utils.py [hunk 5]
```diff
@@ -510,6 +518,65 @@ def _build_index(self, nodes_op):
...
+                if last_cpp.kind() in [TUPLE_CONSTRUCT_KIND, LIST_CONSTRUCT_KIND]:
+                    # we need check if the tensor tuple or tensor list is produced
+                    # by a list/tuple construct node. If so, we can unpack the tuple
+                    # or list manunally.
+                    _logger.debug('List/Tuple Construct Node(cpp) %s', str(last_cpp))
+                    _logger.debug('List/Tuple Unpack Node(cpp) %s', str(unpack_cpp))
+                    assert len(list(unpack_cpp.outputs())) == len(list(last_cpp.inputs()))
```
Comment (QuanluZhang) [inline src/sdk/pynni/nni/_graph_utils.py:558]: a little confused about this assert, i think the main reason is i don't understand what is `last_cpp`

File: src/sdk/pynni/nni/_graph_utils.py (hunk 1 above)
Comment (zheng-ningxin) [reply src/sdk/pynni/nni/_graph_utils.py:207]: Key nodes are the nodes that should not be merged into other nodes. In the past, we only take the `aten::` nodes as the important(key) nodes. In this pr, we also take the list/tuple unpack nodes as the key nodes.

File: src/sdk/pynni/nni/_graph_utils.py (hunk 3 above)
Comment (zheng-ningxin) [reply src/sdk/pynni/nni/_graph_utils.py:534]: Take the shufflenet as an example:
https://github.com/microsoft/nni/issues/2581

File: src/sdk/pynni/nni/_graph_utils.py (hunk 4 above)
Comment (zheng-ningxin) [reply src/sdk/pynni/nni/_graph_utils.py:551]: The `last` node actually refers to the previous(last) visited node.  In most scenarios, this `last_cpp` is the corresponding construct node of the tuple/list.

**Pull request merged**